src/coordinator/api/v1/handler/prometheus/remote/write.go

```diff
@@ -104,11 +121,70 @@ func (h *PromWriteHandler) parseRequest(r *http.Request) (*prompb.WriteRequest,
 }
 
 func (h *PromWriteHandler) write(ctx context.Context, r *prompb.WriteRequest) error {
-	requests := make([]execution.Request, len(r.Timeseries))
-	for idx, t := range r.Timeseries {
-		requests[idx] = newLocalWriteRequest(storage.PromWriteTSToM3(t), h.store)
+	var (
+		wg                   sync.WaitGroup
+		writeUnaggregatedErr error
+		writeAggregatedErr   error
+	)
+	if h.downsampler != nil {
+		// If writing downsampled aggregations, write them async
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+
+			var (
+				metricsAppender = h.downsampler.NewMetricsAppender()
+				multiErr        xerrors.MultiError
+			)
+			for _, ts := range r.Timeseries {
+				metricsAppender.Reset()
+				for _, label := range ts.Labels {
+					metricsAppender.AddTag(label.Name, label.Value)
+				}
+
+				samplesAppender, err := metricsAppender.SamplesAppender()
+				if err != nil {
+					multiErr = multiErr.Add(err)
+					continue
+				}
+
+				for _, elem := range ts.Samples {
+					err := samplesAppender.AppendGaugeSample(elem.Value)
+					if err != nil {
+						multiErr = multiErr.Add(err)
+					}
+				}
+			}
+
+			metricsAppender.Finalize()
+
+			writeAggregatedErr = multiErr.FinalError()
+		}()
 	}
-	return execution.ExecuteParallel(ctx, requests)
+
+	if h.store != nil {
+		// Write the unaggregated points out, don't spawn goroutine
+		// so we reduce number of goroutines just a fraction
+		requests := make([]execution.Request, 0, len(r.Timeseries))
+		for _, t := range r.Timeseries {
+			write := storage.PromWriteTSToM3(t)
+			write.Attributes = storage.Attributes{
+				MetricsType: storage.UnaggregatedMetricsType,
+			}
+			request := newLocalWriteRequest(write, h.store)
+			requests = append(requests, request)
+		}
+		writeUnaggregatedErr = execution.ExecuteParallel(ctx, requests)
```

Comment:
seems like we are using two very different ways of writing aggregated vs unaggregated (multiErr vs failing on first error). If you think multiErr is better, then let's switch to that ?

Reply:
just to clarify: my worry here is that for downsampled metrics we would end up writing all endpoints, ignoring any intermediate errors but for unaggregated metrics, we will fail on first error. I think that behavior should be consistent ? Maybe consider hiding the actual downsampled and unaggregated writes behind another object instead of putting this in remote/write.go ? We also have native/write.go and you don't want to replicate the logic in two places.

Reply:
Sounds good, I've refactored this to always try and write every incoming value regardless of how many errors we encounter.